### PR TITLE
Increase resilience of AsyncGetCallTrace

### PR DESCRIPTION
**Pull request by gacholio**
- Partially inline getJNIMethodID to avoid the assertion case
- Disable trace and assertions during ASGCT
- Build JIT resolve frame on the C stack to avoid writing to the java
stack

[ci skip]

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

**Comment by gacholio** [comment]
If we want to get fancy in the future, a better approach to the trace modifications would be to disable normal tracepoints and change assertions so that they do not print and cause a crash. This would remove the need for inlining getMethodIndex.

File: runtime/j9vm/asgct.cpp [hunk 1]
```diff
@@ -68,8 +68,20 @@ asyncFrameIterator(J9VMThread * currentThread, J9StackWalkState * walkState)
 {
 	ASGCT_CallFrame *frame = (ASGCT_CallFrame*)walkState->userData1;
 	J9Method *method = walkState->method;
-	J9JNIMethodID *methodID = currentThread->javaVM->internalVMFunctions->getJNIMethodID(currentThread, method);
-	Assert_SC_notNull(methodID);
+	J9JavaVM * vm = currentThread->javaVM;
+	J9Class * declaringClass = J9_CLASS_FROM_METHOD(method);
+	UDATA methodIndex = getMethodIndexUnchecked(method);
+	if (UDATA_MAX == methodIndex) {
+		*(UDATA*)UDATA_MAX = UDATA_MAX;
```
Comment (tajila) [inline runtime/j9vm/asgct.cpp:75]: Can we just return `J9_STACKWALK_STOP_ITERATING` instead?

Comment (gacholio) [reply runtime/j9vm/asgct.cpp:75]: That would result in inaccurate traces being returned (which is a risk anyway I suppose). The idea was that only a completely successful stack walk would result in returning frames.

Comment (tajila) [reply runtime/j9vm/asgct.cpp:75]: Understood. Can we make this more clear be defining something like `#define TRIGGER_SIGSEV() *(UDATA*)UDATA_MAX = UDATA_MAX`

And adding a comment stating that `only a completely successful stack walk would result in returning frames`

Comment (gacholio) [reply runtime/j9vm/asgct.cpp]: Done.

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
Jenkins test sanity win jdk8

**Comment by tajila** [comment]
Jenkins test sanity plinux jdk11

File: runtime/j9vm/asgct.cpp [hunk 2]
```diff
@@ -112,19 +146,37 @@ protectedASGCT(J9PortLibrary *portLib, void *arg)
...
-		J9StackWalkState walkState;
+		/* Disable the JIT metadata cache. The cache may be in an inconsistent state, and may
+		 * need to allocate memory (which is undesireable in a signal handler).
+		 */
+		currentThread->jitArtifactSearchCache = (void*)((UDATA)currentThread->jitArtifactSearchCache | J9_STACKWALK_NO_JIT_CACHE);
+		/* Disable trace and assertions. The current thread may be in the process of updating the trace
+		 * buffers when it was interrupted. Assertions during ASGCT are also undesireable, better to let
+		 * the thread continue on to crash, which is handled.
+		 */
+		currentThread->privateFlags2 |= J9_PRIVATE_FLAGS2_ASYNC_GET_CALL_TRACE;
```
Comment (keithc-ca) [inline runtime/j9vm/asgct.cpp:175]: @gacholio Should this flag be reset after the stack walk?

Comment (gacholio) [reply runtime/j9vm/asgct.cpp:175]: It is reset in the caller, along with all the other modified values.

**Pull request merged**